src/main/drivers/bus_i2c_impl.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "platform.h"
+
+#include "drivers/io_types.h"
+#include "drivers/rcc_types.h"
+
+#define I2C_SHORT_TIMEOUT            ((uint32_t)0x1000)
+#define I2C_LONG_TIMEOUT             ((uint32_t)(10 * I2C_SHORT_TIMEOUT))
+#define I2C_DEFAULT_TIMEOUT          I2C_SHORT_TIMEOUT
+
+#define I2C_PIN_SEL_MAX 3
+
+typedef struct i2cHardware_s {
+    I2CDevice device;
+    I2C_TypeDef *reg;
+    ioTag_t sclPins[I2C_PIN_SEL_MAX];
+    ioTag_t sdaPins[I2C_PIN_SEL_MAX];
+    rccPeriphTag_t rcc;
+#if !defined(STM32F303xC)
+    uint8_t ev_irq;
+    uint8_t er_irq;
+#endif
+} i2cHardware_t;
+
+extern const i2cHardware_t i2cHardware[];
+
+#if defined(STM32F1) || defined(STM32F4)
+typedef struct i2cState_s {
+    volatile bool error;
```

Comment:
ATOMIC_READ (implemented as cast to volatile) will explicitly mark that value is read in one place of loop.
But with compilers improving (and -flto), some other member may stay cached in register and not read back.

In multithreaded program, you will use some synchronization primitive (pthread_mutex_lock) that contains memory barrier (both compiler and hardware).
In CF case I'm not sure. This case may be implemented as primitive spinlock ('ISR_WAIT(condition) or `ISR_WAIT(variable)` + timeout variants), but in general some standard way should be used. Using `volatile` leads to hard to read, possibly buggy code that will need major rewrite is RTOS migration is decided in future ...